File: python/2-D Dynamic Programming/regular_expression_matching.py
class Solution:
    """
    Regular Expression Matching
    
    Problem: Given an input string s and a pattern p, implement regular expression 
    matching with support for '.' and '*'.
    - '.' Matches any single character
    - '*' Matches zero or more of the preceding element
    
    Example: s = "aa", p = "a*"
    Output: True (a* matches "aa")
    
    Approach: Dynamic Programming with Memoization
    - Use DFS with memoization to explore all possible matching combinations
    - Handle three cases: exact match, dot match, and star match
    - For star matches, try both zero and one-or-more occurrences
    - Memoize results to avoid recomputation
    - State: (i, j) = can match s[i:] with p[j:]
    
    Time Complexity: O(m * n) where m = len(s), n = len(p)
    Space Complexity: O(m * n) for memoization cache
    """
    
    def isMatch(self, s: str, p: str) -> bool:
        # Memoization cache: (i, j) -> can match s[i:] with p[j:]
        cache = {}

        def dfs(i, j):
            # Check memoization cache first to avoid recomputation
            if (i, j) in cache:
                return cache[(i, j)]
            
            """
            DFS helper to check if s[i:] matches p[j:]
            Args:
                i: current position in string s
                j: current position in pattern p
            Returns:
                True if s[i:] matches p[j:], False otherwise
            """
            # Base case: both strings exhausted
            if i >= len(s) and j >= len(p):
                return True
            # Base case: pattern exhausted but string not
            if j >= len(p):
                return False
            
            # Check if current characters match (exact or dot)
            match = i < len(s) and (s[i] == p[j] or p[j] == ".")
            
            # Handle star pattern (zero or more occurrences)
            if (j + 1) < len(p) and p[j + 1] == "*":
                # Try zero occurrences (skip current pattern) or one-or-more occurrences
                # Cache the result to avoid recomputing this subproblem
                cache[(i, j)] = dfs(i, j + 2) or (match and dfs(i + 1, j))
                return cache[(i, j)]
            
            # Handle exact/dot match (move both pointers)
            if match:
                # Cache the result for exact/dot match
                cache[(i, j)] = dfs(i + 1, j + 1)
                return cache[(i, j)]
            
            # No match possible - cache the result
            cache[(i, j)] = False
            return False
        
        return dfs(0, 0)

# Test cases
def test_regular_expression_matching():
    """Test cases for Regular Expression Matching problem"""
    
    # Test case 1: Basic exact match
    print("Test 1: s='aa', p='a'")
    sol = Solution()
    result = sol.isMatch("aa", "a")
    expected = False  # 'a' doesn't match 'aa'
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 1 passed\n")
    
    # Test case 2: Star pattern
    print("Test 2: s='aa', p='a*'")
    result = sol.isMatch("aa", "a*")
    expected = True  # 'a*' matches 'aa'
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 2 passed\n")
    
    # Test case 3: Dot pattern
    print("Test 3: s='ab', p='a.'")
    result = sol.isMatch("ab", "a.")
    expected = True  # 'a.' matches 'ab'
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 3 passed\n")
    
    # Test case 4: Complex star pattern
    print("Test 4: s='aab', p='c*a*b'")
    result = sol.isMatch("aab", "c*a*b")
    expected = True  # 'c*a*b' matches 'aab'
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 4 passed\n")
    
    # Test case 5: No match
    print("Test 5: s='mississippi', p='mis*is*p*.'")
    result = sol.isMatch("mississippi", "mis*is*p*.")
    expected = False  # Pattern doesn't match
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 5 passed\n")
    
    # Test case 6: Empty strings
    print("Test 6: s='', p=''")
    result = sol.isMatch("", "")
    expected = True  # Empty matches empty
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 6 passed\n")
    
    # Test case 7: Empty string with pattern
    print("Test 7: s='', p='a*'")
    result = sol.isMatch("", "a*")
    expected = True  # 'a*' can match zero 'a's
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 7 passed\n")
    
    # Test case 8: String with empty pattern
    print("Test 8: s='a', p=''")
    result = sol.isMatch("a", "")
    expected = False  # Empty pattern doesn't match non-empty string
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 8 passed\n")
    
    # Test case 9: Multiple stars
    print("Test 9: s='aaa', p='a*a*a*'")
    result = sol.isMatch("aaa", "a*a*a*")
    expected = True  # Multiple stars can match
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 9 passed\n")
    
    # Test case 10: Dot with star
    print("Test 10: s='ab', p='.*'")
    result = sol.isMatch("ab", ".*")
    expected = True  # '.*' matches any string
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 10 passed\n")
    
    # Test case 11: Complex pattern
    print("Test 11: s='ab', p='.*c'")
    result = sol.isMatch("ab", ".*c")
    expected = False  # '.*c' doesn't match 'ab'
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 11 passed\n")
    
    # Test case 12: Edge case with stars
    print("Test 12: s='a', p='ab*'")
    result = sol.isMatch("a", "ab*")
    expected = True  # 'ab*' matches 'a' (b* matches zero b's)
    print(f"Result: {result}, Expected: {expected}")
    assert result == expected
    print("✓ Test 12 passed\n")
    
    print("All tests passed!")


if __name__ == "__main__":
    test_regular_expression_matching()

File: C++/2-D Dynamic Programming/regular_expression_matching.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /**
     * Function to check if string `s` matches pattern `p`
     * where `p` can include '.' (matches any character) and '*' (matches 0 or more of previous char).
     *
     * Approach: DFS + Memoization
     * - Recursively try to match `s[i:]` and `p[j:]`
     * - Memoize intermediate results to avoid recomputation.
     *
     * Time Complexity: O(m * n) where m = s.size(), n = p.size()
     * Space Complexity: O(m * n) for the memoization table
     */
    bool isMatch(string s, string p) {
        int m = s.size(), n = p.size();

        // Memoization table: map from pair<int,int> → bool
        unordered_map<long long, bool> memo;

        // Define the DFS helper function as a lambda
        function<bool(int, int)> dfs = [&](int i, int j) -> bool {
            // Encode (i,j) as a single key
            long long key = ((long long)i << 32) | j;

            // Check if result is already computed
            if (memo.find(key) != memo.end())
                return memo[key];

            // Base case: both strings are exhausted
            if (i >= m && j >= n)
                return memo[key] = true;

            // Base case: pattern exhausted but string remains
            if (j >= n)
                return memo[key] = false;

            // Check if current characters match (or dot)
            bool match = (i < m) && (s[i] == p[j] || p[j] == '.');

            // If next pattern character is '*'
            if ((j + 1) < n && p[j + 1] == '*') {
                // Try zero occurrence of p[j] or one-or-more if match
                memo[key] = dfs(i, j + 2) || (match && dfs(i + 1, j));
                return memo[key];
            }

            // Regular match: advance both if match
            if (match) {
                memo[key] = dfs(i + 1, j + 1);
                return memo[key];
            }

            // No match
            return memo[key] = false;
        };

        return dfs(0, 0);
    }
};

/*
=================================================
Line-by-line C++ syntax explanation:
=================================================

#include <bits/stdc++.h>
    -> Includes all standard C++ libraries (like <vector>, <string>, <unordered_map>, etc.)

using namespace std;
    -> Allows using names from the standard library (like string, vector) without prefixing with std::

class Solution {
public:
    -> Defines a class named Solution with public methods.

bool isMatch(string s, string p) {
    -> Defines a method `isMatch` that takes two strings `s` and `p` and returns a bool.

int m = s.size(), n = p.size();
    -> Get the lengths of `s` and `p`.

unordered_map<long long, bool> memo;
    -> Hash table to store intermediate results.
       Uses `long long` as key (to encode (i,j)) and `bool` as value.

function<bool(int, int)> dfs = [&](int i, int j) -> bool {
    -> Defines a lambda function `dfs` that captures `memo` and `m`, `n` by reference.

long long key = ((long long)i << 32) | j;
    -> Encode the pair (i,j) into a single long long key.

if (memo.find(key) != memo.end())
    return memo[key];
    -> If result for (i,j) already computed, return cached result.

if (i >= m && j >= n)
    return memo[key] = true;
    -> Base case: both strings are fully matched.

if (j >= n)
    return memo[key] = false;
    -> Base case: pattern exhausted but string still has characters.

bool match = (i < m) && (s[i] == p[j] || p[j] == '.');
    -> Check if current characters match (including '.' wildcard).

if ((j + 1) < n && p[j + 1] == '*') {
    -> If next character in pattern is '*'.

memo[key] = dfs(i, j + 2) || (match && dfs(i + 1, j));
    -> Either skip the '*' (zero occurrence) or use it (one/more occurrences).

if (match) {
    memo[key] = dfs(i + 1, j + 1);
    return memo[key];
}
    -> If current characters match, move to next in both strings.

return memo[key] = false;
    -> Otherwise, no match.

return dfs(0, 0);
    -> Start recursion from indices (0,0)

=================================================
Time Complexity:
    - There are O(m * n) unique (i,j) states, each computed once → O(m * n)

Space Complexity:
    - Memoization table stores O(m * n) entries → O(m * n)
=================================================
*/

